uint256.go

```diff
@@ -576,6 +609,38 @@ func (z *Int) SMod(x, y *Int) *Int {
 	return z
 }
 
+// MulModWithReciprocal calculates the modulo-m multiplication of x and y
+// and returns z, using the reciprocal of m provided as the mu parameter.
+// Use uint256.Reciprocal to calculate mu from m.
+// If m == 0, z is set to 0 (OBS: differs from the big.Int)
+func (z *Int) MulModWithReciprocal(x, y, m *Int, mu *[5]uint64) *Int {
+	if x.IsZero() || y.IsZero() || m.IsZero() {
+		return z.Clear()
+	}
+	p := umul(x, y)
+
+	if m[3] != 0 {
+		r := reduce4(p, m, *mu)
+		return z.Set(&r)
+	}
+
+	var (
```

Comment:
It has been modified so that it is limited to m[3] != 0. Hence the first comment inside the function is "Note: specialized for m[3] != 0".
It's not so clear this way, so I'll adjust the comments preceding the function.